src/components/ModelTableList/ModelTable/ModelTable.tsx

```diff
@@ -0,0 +1,208 @@
+import { MainTable } from "@canonical/react-components";
+import type { MainTableRow } from "@canonical/react-components/dist/components/MainTable/MainTable";
+
+import ModelActions from "components/ModelActions";
+import Status from "components/Status";
+import TruncatedTooltip from "components/TruncatedTooltip";
+import { getActiveUsers, getControllerData } from "store/juju/selectors";
+import type { Controllers, ModelData } from "store/juju/types";
+import {
+  extractOwnerName,
+  getModelStatusGroupData,
+} from "store/juju/utils/models";
+import { useAppSelector } from "store/store";
+
+import CloudCell from "../CloudCell";
+import ModelNameCell from "../ModelNameCell";
+import ModelSummary from "../ModelSummary";
+import WarningMessage from "../StatusGroup/WarningMessage";
+import {
+  generateCloudAndRegion,
+  generateTableHeaders,
+  getControllerName,
+  getCredential,
+  getLastUpdated,
+  getRegion,
+} from "../shared";
+import { TestId } from "../types";
+
+/**
+  Returns the model info and statuses in the proper format for the table data.
+  @param  models The list of models
+  @return The formatted table rows.
+*/
+function generateModelTableList(
+  models: ModelData[],
+  activeUsers: Record<string, string>,
+  controllers: Controllers | null,
+  groupBy: "status" | "owner" | "cloud",
+  groupLabel?: string,
+) {
+  const modelsList: MainTableRow[] = [];
+  models.forEach((model) => {
+    const activeUser = activeUsers[model.uuid];
+    const { highestStatus } = getModelStatusGroupData(model);
+    const owner = model.info ? extractOwnerName(model.info["owner-tag"]) : null;
+    const region = getRegion(model);
+    const cloud = <CloudCell model={model} />;
+    const credential = getCredential(model);
+    const controller = getControllerName(model, controllers);
+    const lastUpdated = getLastUpdated(model);
+
+    const columns = [
+      {
+        "data-testid": TestId.COLUMN_NAME,
+        content: (
+          <>
+            <ModelNameCell model={model} />
+            {groupBy === "status" && groupLabel === "blocked" ? (
+              <WarningMessage model={model} />
+            ) : null}
+          </>
+        ),
+      },
+      {
+        "data-testid": TestId.COLUMN_SUMMARY,
+        content: (
+          <ModelSummary
+            modelData={model}
+            ownerTag={model.info?.["owner-tag"]}
+          />
+        ),
+        className: "u-overflow--visible",
+      },
+      // Conditionally include columns based on the group type
+      ...(groupBy !== "owner"
+        ? [
+            {
+              "data-testid": TestId.COLUMN_OWNER,
+              content: (
+                <TruncatedTooltip message={owner}>{owner}</TruncatedTooltip>
+              ),
+            },
+          ]
+        : []),
```

Comment:
I wonder if these would be a little more readable if we were to introduce a utility to conditionally display these. Maybe something like this?

```suggestion
      ...displayForGroups(groupBy, ["cloud", "status"], {
              "data-testid": TestId.COLUMN_OWNER,
              content: (
                <TruncatedTooltip message={owner}>{owner}</TruncatedTooltip>
              ),
            }),
```

What do you think?

Reply:
Done!